adrs/2068541249/README.md

```diff
@@ -0,0 +1,154 @@
+# Configuration of baseUrl and paths of jsconfig/tsconfig files in applications
+
+* **State:** Draft
+* **Created:** 2020-11-18
+* **Tags:** tsconfig,jsconfig,application
+
+## Context
+
+First, let's define the meaning of `application` in this context:
+
+**Applications:** a deployable project that is not used as a library of another
+project.
+
+Current, most JavaScript tools use `<rootDir>/tsconfig.json`, or
+`<rootDir>/jsconfig.json` to help the IDEs or tools to resolve the file locations
+use the following two configurations:
+
+* **compilerOptions.baseUrl:** Base directory to resolve non-relative module
+  names.
+* **compilerOptions.paths:** Specify path mapping to be computed relative to
+  baseUrl option.
+
+For example:
+
+```json
+{
+  "compilerOptions": {
+    "baseUrl": ".",
+    "paths": {
+      "@/*": ["./src/*"]
+    }
+  }
+}
+```
+
+Let's assume the following directory structure as a reference for the examples:
+
+```text
+tree . -L 2
+
+├── tsconfig.json
+├── src
+│   ├── components/...
+│   ├── design/...
+│   └── react-query/...
+│   └── http-client.ts
+```
+
+There are some caveats configuring such files.
+
+For example, using the following configuration:
+
+```json
+{
+  "compilerOptions": {
+    "baseUrl": "src"
+  }
+}
+```
+
+That means, you could import the code as if those files or directories were
+modules:
+
+```tsx
+import { httpClient } from 'http-client';
+import { useMyQuery } from 'react-query';
+```
+
+That configuration brings some critical questions about the application.
+
+How do we know if those imports come from `node_modules` or the imports are
+part of our codebase?
+
+Although it may be simply to answer the question by scanning the directories and
+files under `./src`.
+
+In other cases like `import { useMyQuery } from 'react-query';` where you
+may have `react-query` as a dependency, and you used `./src/react-query`
+directory to store your React Query related things as an example, you need to
+scan the files to make sure that the imports don't belong to a dependency
+package.
```

Comment:
Great point!  In other words, you could inadvertently override an external module locally.

Reply:
I don't know if this story is relevant to Typescript. Still, In the past, I've experienced a scenario where external dependencies had a common name like `http-client` and that 3rd party dependency was resolving to my local `src/http-client/index`.  This is the nightmare scenario.

Reply:
Yes, sir, the error-prune is really high with such configuration.